File: code_snippets/api-guide/model/java/add_images_with_concepts.java
package com.clarifai.example;

import com.clarifai.grpc.api.*;
import com.clarifai.channel.ClarifaiChannel;
import com.clarifai.credentials.ClarifaiCallCredentials;
import com.clarifai.grpc.api.status.StatusCode;

public class ClarifaiExample {

    //////////////////////////////////////////////////////////////////////////////////
    // In this section, we set the user authentication, app ID, and the images and 
    // concepts we want to add. Change these strings to run your own example.
    /////////////////////////////////////////////////////////////////////////////////

    static final String USER_ID = "YOUR_USER_ID_HERE";
    //Your PAT (Personal Access Token) can be found in the portal under Authentication
    static final String PAT = "YOUR_PAT_HERE";
    static final String APP_ID = "YOUR_APP_ID_HERE";
    // Change these to add your own images with concepts
    static final String IMAGE_URL_1 = "https://samples.clarifai.com/puppy.jpeg";
    static final String IMAGE_URL_2 = "https://samples.clarifai.com/wedding.jpg";
    static final String CONCEPT_ID_1 = "charlie";
    static final String CONCEPT_ID_2 = "our_wedding";
    static final String CONCEPT_ID_3 = "our_wedding";
    static final String CONCEPT_ID_4 = "charlie";
    static final String CONCEPT_ID_5 = "cat";

    ///////////////////////////////////////////////////////////////////////////////////
    // YOU DO NOT NEED TO CHANGE ANYTHING BELOW THIS LINE TO RUN THIS EXAMPLE
    ///////////////////////////////////////////////////////////////////////////////////	

    public static void main(String[] args) {

        V2Grpc.V2BlockingStub stub = V2Grpc.newBlockingStub(ClarifaiChannel.INSTANCE.getGrpcChannel())
            .withCallCredentials(new ClarifaiCallCredentials(PAT));

        MultiInputResponse postInputsResponse = stub.postInputs(
            PostInputsRequest.newBuilder()
            .setUserAppId(UserAppIDSet.newBuilder().setUserId(USER_ID).setAppId(APP_ID))
            .addInputs(
                Input.newBuilder()
                .setData(
                    Data.newBuilder()
                    .setImage(
                        Image.newBuilder()
                        .setUrl(IMAGE_URL_1)
                        .setAllowDuplicateUrl(true)
                    )
                    .addConcepts(Concept.newBuilder().setId(CONCEPT_ID_1).setValue(1))
                    .addConcepts(Concept.newBuilder().setId(CONCEPT_ID_2).setValue(0))
                )
            )
            .addInputs(
                Input.newBuilder()
                .setData(
                    Data.newBuilder()
                    .setImage(
                        Image.newBuilder()
                        .setUrl(IMAGE_URL_2)
                        .setAllowDuplicateUrl(true)
                    )
                    .addConcepts(Concept.newBuilder().setId(CONCEPT_ID_3).setValue(1))
                    .addConcepts(Concept.newBuilder().setId(CONCEPT_ID_4).setValue(0))
                    .addConcepts(Concept.newBuilder().setId(CONCEPT_ID_5).setValue(0))
                )
            )
            .build()
        );

        if (postInputsResponse.getStatus().getCode() != StatusCode.SUCCESS) {
            for (Input input: postInputsResponse.getInputsList()) {
                System.out.println("Input " + input.getId() + " status: ");
                System.out.println(input.getStatus() + "\n");
            }

            throw new RuntimeException("Post inputs failed, status: " + postInputsResponse.getStatus());
        }

    }

}

File: code_snippets/api-guide/model/php/add_images_with_concepts.php
<?php

require __DIR__ . "/vendor/autoload.php";

////////////////////////////////////////////////////////////////////////////////////
// In this section, we set the user authentication, app ID, and the images and 
// concepts we want to add. Change these strings to run your own example.
////////////////////////////////////////////////////////////////////////////////////

$USER_ID = "YOUR_USER_ID_HERE";
// Your PAT (Personal Access Token) can be found in the portal under Authentification
$PAT = "YOUR_PAT_HERE";
$APP_ID = "YOUR_APP_ID_HERE";
// Change these to add your own images with concepts
$IMAGE_URL_1 = "https://samples.clarifai.com/puppy.jpeg";
$IMAGE_URL_2 = "https://samples.clarifai.com/wedding.jpg";
$CONCEPT_ID_1 = "charlie";
$CONCEPT_ID_2 = "our_wedding";
$CONCEPT_ID_3 = "our_wedding";
$CONCEPT_ID_4 = "charlie";
$CONCEPT_ID_5 = "cat";

///////////////////////////////////////////////////////////////////////////////////
// YOU DO NOT NEED TO CHANGE ANYTHING BELOW THIS LINE TO RUN THIS EXAMPLE
///////////////////////////////////////////////////////////////////////////////////

use Clarifai\Api\Concept;
use Clarifai\ClarifaiClient;
use Clarifai\Api\Data;
use Clarifai\Api\Image;
use Clarifai\Api\Input;
use Clarifai\Api\PostInputsRequest;
use Clarifai\Api\Status\StatusCode;
use Clarifai\Api\UserAppIDSet;

$client = ClarifaiClient::grpc();

$metadata = ["Authorization" => ["Key " . $PAT]];

$userDataObject = new UserAppIDSet([
    "user_id" => $USER_ID,
    "app_id" => $APP_ID
]);

// Let's make a RPC call to the Clarifai platform. It uses the opened gRPC client channel to communicate a
// request and then wait for the response
[$response, $status] = $client->PostInputs(
    // The request object carries the request along with the request status and other metadata related to the request itself
    new PostInputsRequest([
        "user_app_id" => $userDataObject,
        "inputs" => [
            new Input([
                // The Input object wraps the Data object in order to meet the API specification                    
                "data" => new Data([
                    // The Data object is constructed around the Image object. It offers a container that has additional image independent
                    // metadata. In this particular use case, no other metadata is needed to be specified
                    "image" => new Image([
                        // In the Clarifai platform, an image is defined by a special Image object
                        "url" => $IMAGE_URL_1,
                        "allow_duplicate_url" => true
                    ]),
                    "concepts" => [
                        new Concept([
                            "id" => $CONCEPT_ID_1,
                            "value" => 1
                        ]),
                        new Concept([
                            "id" => $CONCEPT_ID_2,
                            "value" => 0
                        ])
                    ]
                ])
            ]),
            new Input([
                "data" => new Data([
                    "image" => new Image([
                        "url" => $IMAGE_URL_2,
                        "allow_duplicate_url" => true
                    ]),
                    "concepts" => [
                        new Concept([
                            "id" => $CONCEPT_ID_3,
                            "value" => 1
                        ]),
                        new Concept([
                            "id" => $CONCEPT_ID_4,
                            "value" => 0
                        ]),
                        new Concept([
                            "id" => $CONCEPT_ID_5,
                            "value" => 0
                        ])
                    ]
                ])
            ])
        ]
    ]),
    $metadata
)->wait();

// A response is returned and the first thing we do is check the status of it
// A successful response will have a status code of 0; otherwise, there is some error
if ($status->code !== 0) {
    throw new Exception("Error: {$status->details}");
}
// In addition to the RPC response status, there is a Clarifai API status that reports if the operation was a success or failure 
// (not just that the communication was successful)
if ($response->getStatus()->getCode() != StatusCode::SUCCESS) {
    print $response->getStatus()->getDetails();
    throw new Exception("Failure response: " . $response->getStatus()->getDescription());
}

echo $response->serializeToJsonString();
